File: fe/fe-core/src/main/java/com/starrocks/planner/stream/StreamJoinNode.java
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


package com.starrocks.planner.stream;

import com.starrocks.analysis.BinaryPredicate;
import com.starrocks.analysis.Expr;
import com.starrocks.analysis.JoinOperator;
import com.starrocks.analysis.TableRef;
import com.starrocks.planner.JoinNode;
import com.starrocks.planner.PlanNode;
import com.starrocks.planner.PlanNodeId;
import com.starrocks.sql.optimizer.operator.stream.IMTInfo;
import com.starrocks.thrift.TEqJoinCondition;
import com.starrocks.thrift.TExplainLevel;
import com.starrocks.thrift.TPlanNode;
import com.starrocks.thrift.TPlanNodeType;
import com.starrocks.thrift.TStreamJoinNode;
import org.apache.commons.collections4.CollectionUtils;

import java.util.List;
import java.util.stream.Collectors;
import javax.ws.rs.NotSupportedException;

public class StreamJoinNode extends JoinNode {
    // TODO: support bi-stream join
    private IMTInfo rightIMT;

    public StreamJoinNode(PlanNodeId id, PlanNode outer, PlanNode inner, TableRef innerRef, List<Expr> eqJoinConjuncts,
                          List<Expr> otherJoinConjuncts) {
        super("StreamJoin", id, outer, inner, JoinOperator.INNER_JOIN, eqJoinConjuncts, otherJoinConjuncts);
    }

    public StreamJoinNode(PlanNodeId id, PlanNode outer, PlanNode inner, JoinOperator joinOp,
                          List<Expr> eqJoinConjuncts, List<Expr> otherJoinConjuncts) {
        super("StreamJoin", id, outer, inner, joinOp, eqJoinConjuncts, otherJoinConjuncts);
    }

    @Override
    public boolean canUsePipeLine() {
        return getChildren().stream().allMatch(PlanNode::canUsePipeLine);
    }

    @Override
    protected void toThrift(TPlanNode msg) {
        msg.node_type = TPlanNodeType.STREAM_JOIN_NODE;
        msg.stream_join_node = new TStreamJoinNode();
        msg.stream_join_node.join_op = joinOp.toThrift();

        if (CollectionUtils.isNotEmpty(eqJoinConjuncts)) {
            for (BinaryPredicate eqJoinPredicate : eqJoinConjuncts) {
                TEqJoinCondition eqJoinCondition = new TEqJoinCondition(
                        eqJoinPredicate.getChild(0).treeToThrift(), eqJoinPredicate.getChild(1).treeToThrift());
                eqJoinCondition.setOpcode(eqJoinPredicate.getOp().getOpcode());
                msg.stream_join_node.addToEq_join_conjuncts(eqJoinCondition);
            }
        }
        if (CollectionUtils.isNotEmpty(otherJoinConjuncts)) {
            for (Expr e : otherJoinConjuncts) {
                msg.stream_join_node.addToOther_join_conjuncts(e.treeToThrift());
            }
            String sqlJoinPredicate = otherJoinConjuncts.stream().map(Expr::toSql).collect(Collectors.joining(","));
            msg.stream_join_node.setSql_join_predicates(sqlJoinPredicate);
        }
    }

    // TODO support bi-stream join
    public void setLeftIMT(IMTInfo imt) { throw new NotSupportedException("TODO"); }

    public void setRightIMT(IMTInfo imt) { this.rightIMT = imt; }

    @Override
    public boolean canPushDownRuntimeFilter() {
        return false;
    }

    @Override
    protected String getNodeExplainString(String detailPrefix, TExplainLevel detailLevel) {
        String joinStr = super.getNodeExplainString(detailPrefix, detailLevel);
        joinStr += detailPrefix + "rhs_imt: ";
        if (rightIMT != null) {
            joinStr += rightIMT + "\n";
        } else {
            joinStr += "empty\n";
        }
        return joinStr;
    }
}

File: be/src/exec/stream/stream_join_node.cpp
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "exec/stream/stream_join_node.h"

namespace starrocks {

Status StreamJoinNode::init(const TPlanNode& tnode, RuntimeState* state) {
    RETURN_IF_ERROR(ExecNode::init(tnode, state));
    DCHECK(tnode.__isset.stream_join_node);

    _join_op = tnode.stream_join_node.join_op;
    if (_join_op != TJoinOp::INNER_JOIN) {
        return Status::InternalError("StreamJoin only supports InnerJoin for now.");
    }

    if (tnode.stream_join_node.__isset.sql_join_predicates) {
        _sql_join_conjuncts = tnode.stream_join_node.sql_join_predicates;
    }
    auto& eq_join_conjuncts = tnode.stream_join_node.eq_join_conjuncts;
    for (const auto& eq_join_conjunct : eq_join_conjuncts) {
        ExprContext* left = nullptr;
        ExprContext* right = nullptr;
        RETURN_IF_ERROR(Expr::create_expr_tree(_pool, eq_join_conjunct.left, &left, state));
        _probe_expr_ctxs.push_back(left);
        auto* left_expr = left->root();
        assert(left_expr->is_slotref());

        RETURN_IF_ERROR(Expr::create_expr_tree(_pool, eq_join_conjunct.right, &right, state));
        _build_expr_ctxs.push_back(right);
        // auto* right_expr = right->root();
        // assert(right_expr->is_slotref());
        // _join_key_descs.emplace_back(pipeline::LookupJoinKeyDesc{&left_expr->type(),
        //                                                          down_cast<vectorized::ColumnRef*>(left_expr),
        //                                                          down_cast<vectorized::ColumnRef*>(right_expr)});
    }

    // other conjuncts.
    RETURN_IF_ERROR(Expr::create_expr_trees(_pool, tnode.stream_join_node.other_join_conjuncts,
                                            &_other_join_conjunct_ctxs, state));

    // TODO: support output columns later.
    if (tnode.stream_join_node.__isset.output_columns) {
        _output_slots.insert(tnode.stream_join_node.output_columns.begin(),
                             tnode.stream_join_node.output_columns.end());
    }

    return Status::OK();
}

pipeline::OpFactories StreamJoinNode::decompose_to_pipeline(pipeline::PipelineBuilderContext* context) {
    OpFactories left_ops = _children[0]->decompose_to_pipeline(context);
    OpFactories right_ops = _children[1]->decompose_to_pipeline(context);
    _left_row_desc = child(0)->row_desc();
    _right_row_desc = child(1)->row_desc();
    // assert right_op must be index_seek operator.
    assert(right_ops.size() >= 1);

    // if (typeid(*(right_ops[0])) != typeid(pipeline::LookupJoinSeekOperatorFactory)) {
    //     throw std::runtime_error("Only support right table as lookup-join table");
    // }

    // auto* right_source = down_cast<SourceOperatorFactory*>(right_ops[0].get());
    // auto* left_source = down_cast<SourceOperatorFactory*>(left_ops[0].get());

    // pipeline::LookupJoinContextParams params(left_source->degree_of_parallelism(),
    //                                          right_source->degree_of_parallelism(), _join_key_descs, _left_row_desc,
    //                                          _right_row_desc, _other_join_conjunct_ctxs);
    // _lookup_join_context = std::make_shared<pipeline::LookupJoinContext>(std::move(params));
    // // Left side
    // auto left_factory = std::make_shared<pipeline::LookupJoinProbeOperatorFactory>(context->next_operator_id(), id(),
    //                                                                                _lookup_join_context);
    // left_ops.emplace_back(std::move(left_factory));
    // context->add_pipeline(left_ops);

    // // TODO: Convert OlapTableScan to IndexSeek here.
    // auto index_seek_factory = down_cast<pipeline::LookupJoinSeekOperatorFactory*>(right_source);
    // index_seek_factory->with_lookup_join_context(_lookup_join_context);
    return right_ops;
}

} // namespace starrocks
